### ISSUE TITLE
Channel API

### ISSUE THREAD

**Issue by martinschroeder** [concurrent-php/ext-async#49]
I want to implement channels (inspired by Go) as a means of exchanging messages between async tasks. The basic concept is comparable to amphp's `Emitter`. A `Channel` is a source of messages, a `Receiver` is a consumer of messages.
```php
namespace Concurrent\Channel;

final class Channel
{
    public function __construct(int $bufferSize = 0) { }
    public function close(?\Throwable $e = null): void { } 
    public function send($value): void { }
    public function receiver(): Receiver { }
}

final class Receiver implements \Iterator
{
    public function close(): void { }
    public function receive($eof = null): mixed { }
}

class ChannelClosedException extends \Exception { }
```
Sending to a channel suspends the sending task until all receivers consumed the new message. A channel can have a buffer that allows to push messages into the channel without blocking as long as there is enough space in the buffer.

Receivers do implement `Iterator` (or `IteratorAggregate`) to allow for consumption of messages using a `foreach` loop (or any other API that works with iterators). Receiving a message suspends the receiving task until a new message sent into the channel. A receiver might receive a message without being suspended if the message is queued in the channel's buffer.
```php
$channel = new Channel();

async (function (Receiver $it) {
    foreach ($it as $v) {
        var_dump($v);
    }
})($channel->receiver());

$channel->send(1);
$channel->send(2);
$channel->close();
```
Simple (keyword-based) example of how the channel API could be used. Implementing this in the extension would allow for a very resource- and memory-effient red (subscription) model. It can also be used to implement a non-blocking multi-channel reader (see Go's `select` statement).

**Comment by kelunik** [comment]
I would start implementing all these things in userland and only start the C implementation once we know what we really want. There's no need for memory efficiency at the moment, it needs for work first and we have to be 100% fine with the API.

**Comment by martinschroeder** [comment]
Most of the channel API can actually be implemented in userland, however the (very useful) `select` use case cannot be implemented in an efficient way. It would require one `Task` for each input channel and support for cancellation. It would also be very wasteful when only reading the next message. An implementation in C can make use of a much more efficient continuation mechanism that comes at almost no overhead.

This API is not something i will be working on in the next weeks. I just wanted to outline my ideas and the direction i am going with this. If you read up a bit on Go‘s channels there is also the non-blocking `select` which is very interesting as well and should be implemented in C for the same reasons related to performance.

**Comment by joelwurtz** [comment]
BTW Regarding channels what would be the userland API ? do you intend to add specific keywords ? Or should it be only a function call like : 

```php
Concurrent\Channel::select(
     Concurrent\Channel::case($channel1, function($channel1Data) {
         // do something when channel1 is received with data from $channel1Data
     }),
     Concurrent\Channel::case($channel2, function($channel2Data) {
         // do something when channel2 is received with data from $channel2Data
     }),
     Concurrent\Channel::defaultCase(function() {
         // no activity
     }) 
)
```

**Comment by martinschroeder** [comment]
I think a keyword-based solution would be great, something like this:
```php
select ($x) {
    case $channel1:
        // Received entry from first channel is accessible from $x
        break;
    case $channel2:
        // Message from second channel stored in $x
        break;
    default:
        // No channel was ready, $x contains NULL.
}
```
This is very much like what golang does, short and readable. Unfortunately it cannot be implemented in an extension due to the `select` keyword. It will likely need something like @joelwurtz propesed API to implement it in the extension.

I am working on code cleanup for now, filesystem IO and support for named pipes / unix domain sockets are up next. It will take some time before i look into channels...

**Comment by martinschroeder** [comment]
Channels have been implemented in #98.
